### PR TITLE
Removing without_storage_info from scored-pool pallet.

### PR DESCRIPTION
**Pull request by hbulgarini**
Removing macro `#[pallet::without_storage_info]` for scored-pool pallet.

### CODE REVIEW

**Comment by ggwpez** [comment]
bot rebase

File: frame/scored-pool/src/lib.rs [hunk 1]
```diff
@@ -264,10 +276,10 @@ pub mod pallet {
 			// Sorts the `Pool` by score in a descending order. Entities which
 			// have a score of `None` are sorted to the beginning of the vec.
 			pool.sort_by_key(|(_, maybe_score)| Reverse(maybe_score.unwrap_or_default()));
-
+			let pool_bounded: PoolT<T, I> = BoundedVec::truncate_from(pool);
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:279]: Not quite happy with the truncating, should be resolved after https://github.com/paritytech/substrate/pull/11998.  
Hopefully gets merged quickly

File: frame/scored-pool/src/lib.rs [hunk 2]
```diff
@@ -307,7 +319,8 @@ pub mod pallet {
 
 			// can be inserted as last element in pool, since entities with
 			// `None` are always sorted to the end.
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:321]: The comment above the `sort_by_key` reads:  
```rust
// Sorts the `Pool` by score in a descending order. Entities which
// have a score of `None` are sorted to the beginning of the vec.
```
so one of the comments must be wrong.

File: frame/scored-pool/src/lib.rs [hunk 3]
```diff
@@ -423,23 +436,27 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
 			.into_iter()
 			.filter(|(_, score)| score.is_some())
 			.take(count as usize)
 			.map(|(account_id, _)| account_id)
 			.collect();
-		new_members.sort();
 
-		let old_members = <Members<T, I>>::get();
-		<Members<T, I>>::put(&new_members);
+		let mut new_members_bounded: MembersT<T, I> = BoundedVec::truncate_from(new_members);
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:449]: Please add a comment why truncating is fine here.

File: frame/scored-pool/src/lib.rs [hunk 4]
```diff
@@ -129,14 +135,17 @@ pub mod pallet {
...
 	pub struct Pallet<T, I = ()>(_);
 
 	#[pallet::config]
 	pub trait Config<I: 'static = ()>: frame_system::Config {
 		/// The currency used for deposits.
 		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
 
+		/// Maximum members length allowed.
+		#[pallet::constant]
+		type MaximumMembers: Get<u32>;
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:147]: Not sure how important it is; but `MemberCount` should not exceed this number.  
Maybe worth adding an integrity check?

File: frame/scored-pool/src/lib.rs [hunk 5]
```diff
@@ -423,23 +436,27 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	/// type function to invoke at the end of the method.
 	fn refresh_members(pool: PoolT<T, I>, notify: ChangeReceiver) {
 		let count = MemberCount::<T, I>::get();
+		let old_members = <Members<T, I>>::get();
+		let pool = pool.into_inner();
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:440]:
```suggestion
```

The `Deref` trait handles this.

**Review by ggwpez — COMMENTED** [review]
Just some comments :+1:

File: frame/scored-pool/src/lib.rs (hunk 3 above)
Comment (ggwpez) [reply frame/scored-pool/src/lib.rs:449]: I think its fine to use here if we add the integrity check that `MemberCount` does not exceed `MaximumMembers`.  
In any case; using `truncate_from` can silently ignore values, so its a good idea to add a comment in the code why it is being used.

File: frame/scored-pool/src/lib.rs (hunk 1 above)
Comment (ggwpez) [reply frame/scored-pool/src/lib.rs:279]: You can approve it if you like, then we can merge it :smile:

File: frame/scored-pool/src/lib.rs [hunk 6]
```diff
@@ -409,7 +422,7 @@ pub mod pallet {
 		#[pallet::weight(0)]
 		pub fn change_member_count(origin: OriginFor<T>, count: u32) -> DispatchResult {
 			ensure_root(origin)?;
-			MemberCount::<T, I>::put(&count);
+			Self::update_member_count(count).map_err(|_| Error::<T, I>::TooManyMembers)?;
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:425]:
```suggestion
			Self::update_member_count(count)?;
```
Isnt this already returning the correct error type?

File: frame/scored-pool/src/lib.rs [hunk 7]
```diff
@@ -485,4 +503,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 		Ok(())
 	}
+
+	/// Make sure the new member count value does not exceed the MaximumMembers
+	fn update_member_count(new_member_count: u32) -> Result<(), Error<T, I>> {
+		ensure!(new_member_count < T::MaximumMembers::get(), Error::<T, I>::TooManyMembers);
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:509]: Why a `<` and not `<=`?

File: frame/scored-pool/src/mock.rs [hunk 8]
```diff
@@ -133,25 +140,24 @@ impl Config for Test {
...
-			(5, 500_000),
-			(10, 500_000),
-			(15, 500_000),
-			(20, 500_000),
-			(31, 500_000),
-			(40, 500_000),
-			(99, 1),
-		],
+	let mut balances = vec![];
+	for i in 1..30 {
```
Comment (ggwpez) [inline frame/scored-pool/src/mock.rs:149]: Does not seem to make a difference, but Rust ranges exclude the upper value.

File: frame/scored-pool/src/tests.rs [hunk 9]
```diff
@@ -297,3 +297,23 @@ fn candidacy_resubmitting_works() {
...
+			assert_ok!(ScoredPool::submit_candidacy(Origin::signed(who)));
+			let index = find_in_pool(who).expect("entity must be in pool") as u32;
+			assert_ok!(ScoredPool::score(Origin::signed(ScoreOrigin::get()), who, index, 99));
+		}
+
+		let submit_candidacy_call =
+			mock::Call::ScoredPool(crate::Call::<Test>::submit_candidacy {});
+
+		assert_noop!(
+			submit_candidacy_call.dispatch(Origin::signed(8)),
```
Comment (ggwpez) [inline frame/scored-pool/src/tests.rs:315]: https://github.com/paritytech/substrate/pull/11927 should make this superfluous; you can directly call the function without wrapping it.

**Review by ggwpez — APPROVED** [review]
Looks good to me, just some nits :+1:

File: frame/scored-pool/src/mock.rs [hunk 10]
```diff
@@ -97,7 +97,9 @@ impl pallet_balances::Config for Test {
 }
 
 thread_local! {
-	pub static MEMBERS: RefCell<Vec<u64>> = RefCell::new(vec![]);
+
+	pub static MEMBERS: RefCell<BoundedVec<u64,ConstU32<10_u32>>> = RefCell::new(bounded_vec![0,10]);
+
```
Comment (kianenigma) [inline frame/scored-pool/src/mock.rs:102]:
```suggestion
	pub static MEMBERS: RefCell<BoundedVec<u64,ConstU32<10_u32>>> = RefCell::new(bounded_vec![0,10]);
```

Comment (kianenigma) [reply frame/scored-pool/src/mock.rs:102]: Or better, ues: 

```
parameter_types! {
    pub static Members: BoundedVec<u64,ConstU32<10_u32>> = bounded_vec![0,10]
}
```

and then you get `Members::get()` and `Members::set()` with the same functionality.

File: frame/scored-pool/src/lib.rs [hunk 11]
```diff
@@ -409,7 +422,7 @@ pub mod pallet {
 		#[pallet::weight(0)]
 		pub fn change_member_count(origin: OriginFor<T>, count: u32) -> DispatchResult {
 			ensure_root(origin)?;
-			MemberCount::<T, I>::put(&count);
+			Self::update_member_count(count)?;
 			Ok(())
```
Comment (ggwpez) [inline frame/scored-pool/src/lib.rs:426]:
```suggestion
			Self::update_member_count(count)
```

There is one unused `std::cell::RefCell` import in the mock file which makes the CI unhappy.

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
@kianenigma can you confirm that this is not live and we do not need audit?

**Review by kianenigma — APPROVED** [review]
Indeed not live in our deployments.

File: frame/scored-pool/src/tests.rs [hunk 12]
```diff
@@ -297,3 +297,20 @@ fn candidacy_resubmitting_works() {
 		assert_eq!(ScoredPool::candidate_exists(who), true);
 	});
 }
+
+#[test]
+fn pool_candidates_exceded() {
```
Comment (ggwpez) [inline frame/scored-pool/src/tests.rs:302]:
```suggestion
fn pool_candidates_exceeded() {
```

File: frame/scored-pool/src/tests.rs [hunk 13]
```diff
@@ -297,3 +297,20 @@ fn candidacy_resubmitting_works() {
...
+	new_test_ext().execute_with(|| {
+		for i in [1, 2, 3, 4, 6] {
```
Comment (ggwpez) [inline frame/scored-pool/src/tests.rs:304]: The `new_test_ext` adds the scores in the wrong order, putting `None` first.

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
Yea looks like it is good to merge. Just need to resolve one conflict :+1:

**Pull request merged**